### PR TITLE
Allow all probes and lifecycle for restartable init containers

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This is a follow-up PR for https://github.com/kubernetes/kubernetes/pull/116429
- Allow restartable init containers to have readiness/liveness probes and lifecycle
- Generate `ContainersReadyCondition` with restartable init containers
- Add probes and lifecycle hooks e2e tests for restartable init containers

#### Which issue(s) this PR fixes:

Fixes #
depends on https://github.com/kubernetes/kubernetes/pull/119179
ref: https://github.com/kubernetes/kubernetes/issues/115934

#### Does this PR introduce a user-facing change?
```release-note
none
```

wrote "none", as we already added a release note in https://github.com/kubernetes/kubernetes/pull/116429.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers
```

### CODE REVIEW

**Review by Tusenka — APPROVED** [review]

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -2816,6 +2816,45 @@ func validatePodResourceClaimSource(claimSource core.ClaimSource, fldPath *field
...
 }
 
+func validateLivenessProbe(probe *core.Probe, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if probe == nil {
+		return allErrs
+	}
+	allErrs = append(allErrs, validateProbe(probe, fldPath)...)
+	if probe.SuccessThreshold != 1 {
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:2826]: I don't remember liveness probe has any limitations on success threshold. Only startup probes has this strange limitation

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -2816,6 +2816,45 @@ func validatePodResourceClaimSource(claimSource core.ClaimSource, fldPath *field
...
+
+func validateStartupProbe(probe *core.Probe, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if probe == nil {
+		return allErrs
+	}
+	allErrs = append(allErrs, validateProbe(probe, fldPath)...)
+	if probe.SuccessThreshold != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("successThreshold"), probe.SuccessThreshold, "must not 1"))
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:2853]:
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath.Child("successThreshold"), probe.SuccessThreshold, "must be 1"))
```

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (SergeyKanzhelev) [reply pkg/apis/core/validation/validation.go:2826]: taking it back. Strange

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -2816,6 +2816,45 @@ func validatePodResourceClaimSource(claimSource core.ClaimSource, fldPath *field
 	return allErrs
 }
 
+func validateLivenessProbe(probe *core.Probe, fldPath *field.Path) field.ErrorList {
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:2819]: would be great to add e2e tests for sidecars with liveness, readiness, and startup probe. For liveness and readiness those can be added to all other readiness and liveness tests we already have.

Comment (gjkim42) [reply pkg/apis/core/validation/validation.go:2819]: Maybe, we can just use `pull-kubernetes-e2e-gce-cos-alpha-features`.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -3202,22 +3241,12 @@ func validateInitContainers(containers []core.Container, regularContainers []cor
...
-			if ctr.StartupProbe != nil && ctr.StartupProbe.SuccessThreshold != 1 {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("startupProbe", "successThreshold"), ctr.StartupProbe.SuccessThreshold, "must be 1"))
+				allErrs = append(allErrs, validateLifecycle(ctr.Lifecycle, idxPath.Child("lifecycle"))...)
 			}
+			allErrs = append(allErrs, validateLivenessProbe(ctr.LivenessProbe, idxPath.Child("livenessProbe"))...)
+			allErrs = append(allErrs, validateReadinessProbe(ctr.ReadinessProbe, idxPath.Child("readinessProbe"))...)
+			allErrs = append(allErrs, validateStartupProbe(ctr.StartupProbe, idxPath.Child("startupProbe"))...)
 
 		default:
 			// These fields are disallowed for init containers.
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:3252]: nit: can you update the error message for non-sidecar container. Instead of `may not be set for init containers` something like `may not be set for init containers without restartPolicy=Always`

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 5]
```diff
@@ -931,6 +931,28 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 			}
 
 			if types.IsRestartableInitContainer(container) {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go:933]: we will also need to figure out whether we need to call `internalLifecycle.PostStopContainer` for restartable init contianers. I think it is needed for topology manager. We may do it as a separate PR, just need to track it somehow

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

LGTM, except the comment about `internalLifecycle.PostStopContainer`. Likely  need to be tracked separately

File: pkg/apis/core/validation/validation.go (hunk 4 above)
Comment (gjkim42) [reply pkg/apis/core/validation/validation.go:3252]: done.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 5 above)
Comment (gjkim42) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:933]: I guess it is already called by `removeContainer`.

**Comment by gjkim42** [comment]
- added `TestLifeCycleHookForRestartableInitContainer` and e2e tests for restartable init container with lifecycle hook

/assign @thockin @mrunalp 

for API and node approval

**Review by thockin — COMMENTED** [review]
Broad strokes this seems OK.

**Review by mrunalp — APPROVED** [review]

**Comment by thockin** [comment]
I have not been tracking this one - is it hoping to merge for 1.28?

**Comment by gjkim42** [comment]
> I have not been tracking this one - is it hoping to merge for 1.28?

Yes, it is targeted at 1.28.

**Review by thockin — COMMENTED** [review]
/approve

for validation

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 6]
```diff
@@ -931,6 +931,28 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
...
+							// restart it.
+							changes.ContainersToKill[status.ID] = containerToKillInfo{
+								name:      container.Name,
+								container: container,
+								message:   fmt.Sprintf("Init container %s failed liveness probe", container.Name),
+								reason:    reasonLivenessProbe,
+							}
+							changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+						}
+						break
```
[the remark this answers is not in the document]
Comment (gjkim42) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:952]: > nit: Would you add a comment explaining this break as well? I spent a few minutes looking at the two above if statements and this break and it wasn't immediately obvious what the rationale is. It's probably right, I just wasn't able to figure it out just from looking at the code..

I think the current code is slightly wrong...

A restartable init container doesn't have to take into account its liveness probe when it determines to start the next init container.

will fix it soon.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gjkim42) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: fixed kubelet to not take into account the container's livenessProbe when it decides to start the next init container.

https://github.com/kubernetes/kubernetes/compare/8ca77dc4bef1be820137a667ebe082e3721cf7c3..d029f0ef0603256c89617c1b4e7898d2f2f87951

File: test/e2e/common/node/lifecycle_hook.go [hunk 7]
```diff
@@ -253,6 +253,254 @@ var _ = SIGDescribe("Container Lifecycle Hook", func() {
 	})
 })
 
+var _ = SIGDescribe("[NodeAlphaFeature:SidecarContainers][Feature:SidecarContainers] Restartable Init Container Lifecycle Hook", func() {
```
Comment (liggitt) [inline test/e2e/common/node/lifecycle_hook.go:256]: these tests have been flaking, with at least one of them failing in about 50% of the runs since they merged:

https://testgrid.k8s.io/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-gce-cos-alpha-features&include-filter-by-regex=NodeAlphaFeature:SidecarContainers|Test&width=20&show-stale-tests=

**Comment by SergeyKanzhelev** [comment]
sidecar-specific job `pull-kubernetes-node-e2e-containerd-sidecar-containers` passes and green. Maybe some interference between jobs
